python/hail/typ.py

```diff
@@ -878,31 +879,32 @@ def _from_java(cls, jtype):
     def _typecheck(self, annotation):
         return
 
-__singletons = {}
-_singleton_classes = {'is.hail.expr.TInt32Optional$': (TInt32, False),
-                      'is.hail.expr.TInt32Required$': (TInt32, True),
-                      'is.hail.expr.TInt64Optional$': (TInt64, False),
-                      'is.hail.expr.TInt64Required$': (TInt64, True),
-                      'is.hail.expr.TFloat32Optional$': (TFloat32, False),
-                      'is.hail.expr.TFloat32Required$': (TFloat32, True),
-                      'is.hail.expr.TFloat64Optional$': (TFloat64, False),
-                      'is.hail.expr.TFloat64Required$': (TFloat64, True),
-                      'is.hail.expr.TBooleanOptional$': (TBoolean, False),
-                      'is.hail.expr.TBooleanRequired$': (TBoolean, True),
-                      'is.hail.expr.TStringOptional$': (TString, False),
-                      'is.hail.expr.TStringRequired$': (TString, True),
-                      'is.hail.expr.TAltAlleleOptional$': (TAltAllele, False),
-                      'is.hail.expr.TAltAlleleRequired$': (TAltAllele, True),
-                      'is.hail.expr.TGenotypeOptional$': (TGenotype, False),
-                      'is.hail.expr.TGenotypeRequired$': (TGenotype, True),
-                      'is.hail.expr.TCallOptional$': (TCall, False),
-                      'is.hail.expr.TCallRequired$': (TCall, True)}
-def __singletons__(scala_type):
-    global __singletons
-    if scala_type not in __singletons:
-        type, required = _singleton_classes[scala_type]
-        __singletons[scala_type] = type(required)
-    return __singletons[scala_type]
+__interns = None
```

Comment:
I think that this (`__interns` and `__interns__(scala)`) now duplicates a lot of the work that's being done by the Intern superclass is doing in `__call__`. Would it be worth changing it back to something like:

```
class Type()
def _from_java:
    ...
    if class_name in _intern_classes:
        return __interns__(class_name)[0].__call__(__interns__(class_name)[1])
    ...
...
__interns__ = {scala_object: (python_class, req)}
```
or does that not do what I think it does/is it ok as is?

Reply:
Good point!  Fixed.  Much cleaner.